src/main/java/jenkins/scm/api/SCMEvent.java

```diff
@@ -247,14 +304,79 @@ public int hashCode() {
     @Override
     public String toString() {
         return String.format(
-                "SCMEvent{type=%s, timestamp=%tc, payload=%s}",
+                "SCMEvent{type=%s, timestamp=%tc, payload=%s, origin=%s}",
                 type,
                 timestamp,
-                payload
+                payload,
+                StringUtils.defaultIfBlank(origin, ORIGIN_UNKNOWN)
         );
     }
 
     /**
+     * Helper method to get the origin of an event from a {@link StaplerRequest}.
+     *
+     * @param req the {@link StaplerRequest}.
+     * @return the origin of the event.
+     */
+    @CheckForNull
+    public static String originOf(@CheckForNull StaplerRequest req) {
+        if (req == null) {
+            return null;
+        }
+        String last = null;
+        StringBuilder result = new StringBuilder();
+        // TODO RFC 7239 support once standard is approved
+        String header = req.getHeader("X-Forwarded-For");
+        if (StringUtils.isNotBlank(header)) {
+            for (String remote : header.split("(,\\s*)")) {
+                if (StringUtils.isBlank(remote)) {
+                    continue;
+                }
+                if (last != null) {
+                    result.append(" => ");
+                }
+                last = StringUtils.trim(remote);
+                result.append(last);
+            }
+        }
+        String remoteHost = req.getRemoteHost();
+        String remoteAddr = req.getRemoteAddr();
+        if (last == null || (!(StringUtils.equals(last, remoteHost) || StringUtils.equals(last, remoteAddr)))) {
+            if (last != null) {
+                result.append(" => ");
+            }
+            if (!StringUtils.isBlank(remoteHost) && !remoteHost.equals(remoteAddr)) {
+                result.append(remoteHost);
+                result.append('/');
+            }
+            result.append(remoteAddr);
+        }
+        result.append(" => ");
+        String scheme = StringUtils.defaultIfBlank(req.getHeader("X-Forwarded-Proto"), req.getScheme());
+        result.append(scheme);
+        result.append("://");
+        result.append(req.getServerName());
+        String portStr = req.getHeader("X-Forwarded-Port");
+        int port;
+        if (portStr != null) {
+            try {
+                port = Integer.parseInt(portStr);
+            } catch (NumberFormatException e) {
+                port = req.getRemotePort();
+            }
+        } else {
+            port = req.getRemotePort();
+        }
+        if (!("http".equals(scheme) && port == 80 || "https".equals(scheme) && port == 443)) {
+            result.append(':');
+            result.append(port);
```

Comment:
Concur with Jesse on both points.  I can follow what you're doing here but would seriously look at the Apache and Guava dependencies to see if we have a utility that can cut out some of this nastiness.